Debug stream debugfs

jsarha opened the pull request:
This commit simply maps the identified debug-stream debug window slot as a debugfs file.

Commit:
ASoC: SOF: IPC4: Add debug-stream debug slot type

Add new debug window slot type for debug-stream protocol. For details see
src/debug/debug_stream/debug_stream_slot.h under SOF sources [1].

[1] https://github.com/thesofproject/sof

Signed-off-by: Jyri Sarha <jyri.sarha@linux.intel.com>

plbossart commented on sound/soc/sof/ipc4-debug-stream.c:
```diff
@@ -0,0 +1,80 @@
+// This file is provided under a dual BSD/GPLv2 license.  When using or
+// redistributing this file, you may do so under either license.
+//
+// Copyright(c) 2024 Intel Corporation. All rights reserved.
```
remove All rights reserved.

plbossart commented on sound/soc/sof/ipc4-debug-stream.c:
```diff
@@ -0,0 +1,80 @@
+// SPDX-License-Identifier: (GPL-2.0-only OR BSD-3-Clause)
```
typo in commit body "encodeing" -> "encoding"

plbossart commented on sound/soc/sof/ipc4-debug-stream.h:
```diff
@@ -0,0 +1,14 @@
+ * This file is provided under a dual BSD/GPLv2 license.  When using or
+ * redistributing this file, you may do so under either license.
+ *
+ * Copyright(c) 2024 Intel Corporation. All rights reserved.
```
remove All rights reserved

plbossart commented on sound/soc/sof/ipc4-debug-stream.c:
```diff
@@ -0,0 +1,80 @@
+	list_add(&dfse->list, &sdev->dfsentry_list);
+
+	debugfs_create_file("debug_stream", 0444, sdev->debugfs_root, dfse, &sof_debug_stream_fops);
+}
```
do we need a new file and prototype for a single function?

ujfalusi commented on sound/soc/sof/ipc4-debug-stream.c:
```diff
@@ -0,0 +1,80 @@
+	if (!buf)
+		return -ENOMEM;
+
+	sof_mailbox_read(sdev, offset, buf, SOF_IPC4_DEBUG_SLOT_SIZE);
```
Would not make sense to skip the magic number as the telemetry does?
If so, then probably creating a single function to get the content of a debug slot by type and use that in both?

jsarha replied:
Sorry @ujfalusi , I do not follow. What magic number you are referring to?

About unifying the telemetry and debug-stream code I agree on general level, but there is the exception part in the current telemetry code, that I do not fully comprehend. And then telemetry debugfs does not support seek and there are some other subtle differences. To unify the code I should know how telemetry debugfs file is used, so that I can verify it works with code borrowed from debug-stream -side. The other way around wont work, as my python debug-stream client disagrees with the telemetry debugfs implementation.

ujfalusi replied:
@jsarha, the first u32 in a slot is the magic number or tag or slot type, it is SOF_IPC4_DEBUG_SLOT_DEBUG_STREAM in this case. For telemetry it is SOF_IPC4_DEBUG_SLOT_TELEMETRY.
The real data is after this magic number.

The `sof_ipc4_query_exception_address()` is just wrapping the `sof_ipc4_find_debug_slot_offset_by_type()` call and skips the magic number to get a pointer for the slot data.
The telemetry.c uses memcpy_fromio() instead sof_mailbox_read(), they do the same at the end.

If I look at the two files from a distance, they are 99% identical with the only difference is:
SOF_IPC4_DEBUG_SLOT_DEBUG_STREAM vs SOF_IPC4_DEBUG_SLOT_TELEMETRY
"debug_stream" vs "exception"

The only tricky part is how to know which slot you need to return, for which the `debug-dsp-ops.c` file gives an answer.

jsarha replied:
Yes, I know how the debug slots work, and I understand the basic structure, but while I was implementing debug-stream, I started off with how telemetry debugfs was implemented, but needed to change couple of things  to get it to work.

Only thing is, that I would like to test the telemetry still works, if I unify the debugfs behind a single function that follows the debug-stream implementation... but I have no idea how telemetry should be used and I can not find any client etc. But yes, maybe I just do it and hope the possible issues will come up in review.

jsarha replied:
> @jsarha, the first u32 in a slot is the magic number or tag or slot type, it is SOF_IPC4_DEBUG_SLOT_DEBUG_STREAM in this case. For telemetry it is SOF_IPC4_DEBUG_SLOT_TELEMETRY. The real data is after this magic number.
> 

Oh, sorry @ujfalusi I read hastily. This is not how the debug slots work. The slot allocation is done in the descriptor slot:

https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/intel/intel_adsp/common/include/adsp_debug_window.h

And the whole allocated 4k slot can be used freely, no need to skip any words. And I rather not skip any words in the generic debug slot debufs mapping, just because its required in telemetry slot when its used for exception dumping.

Maybe I could parameterize it, thou that would look funny too. It would be really helpful to know how (with what tool?) the exception file is used. I'd rather skip the first word in the exception interpreting tool than in in the generic debug slot debugfs implementation.

ujfalusi replied:
Right, yes, you are right. The telemetry slot itself can be split up to different sections and internally it is using 'separator magic' numbers for identification (in theory, but that's another story).
So, here, the ipc4-telemetry.c is exposing the 'exception' section of the telemetry slot, which I guess is the only one present when an exception happens and it will overtake the whole telemetry slot.
This is the reason why the first u32 is skipped I guess, it is a marker in the scope of the telemetry slot.

ujfalusi commented on sound/soc/sof/ipc4-debug-stream.h:
```diff
@@ -0,0 +1,14 @@
+#ifndef __SOUND_SOC_SOF_IPC4_DEBUG_STREAM_H
+#define __SOUND_SOC_SOF_IPC4_DEBUG_STREAM_H
+
+void sof_ipc4_create_debug_stream_debugfs_node(struct snd_sof_dev *sdev);
```
Since this is just a single function, just add it tot he ipc4-priv.h

ujfalusi reviewed (changes requested):
@jsarha, what I would do is:
rename the `ipc4-telemetry.c` to `ipc4-debugfs.c` or something generic and use this to export the raw slots on the needed base.
You can have a single exported function as entry point, you can even move here the `sof_ipc4_find_debug_slot_offset_by_type()` function...

I know, mtrace is also uses debug slot, but it has been special cased out...

The `ipc4-telemetry.h` should remain as it is.

Commit:
ASoC: SOF: ipc4: Add sof_ipc4_create_debug_slot_ro_debugfs_node()

Adds sof_ipc4_create_debug_slot_ro_debugfs_node() -function for
mapping a SOF Intel ipc4 debug window slot as debugfs file. The actual
slot is specified with slot_type parameter. The alternatives are
defined in include/sound/sof/ipc4/header.h and the slot is found with
sof_ipc4_find_debug_slot_offset_by_type(). It also takes the
data_offset parameter that specifies where the payload data in the
slot begins. The portion that is mapped to the debugfs file is
everything after the offset. The last parameter is the name of the
file.

Signed-off-by: Jyri Sarha <jyri.sarha@linux.intel.com>

Commit:
ASoC: SOF: ipc4: Maps debug-stream debug window slot as ro debugfs file

Maps debug-stream debug window slot as read-only debugfs file with
sof_ipc4_create_debug_slot_ro_debugfs_node() -function.

Signed-off-by: Jyri Sarha <jyri.sarha@linux.intel.com>

Commit:
ASoC: SOF: ipc4: Use generic function for mapping "exception" debugfs file

Remove sof_ipc4_create_exception_debugfs_node() and ipc4-telemetry.c,
and use sof_ipc4_create_debug_slot_ro_debugfs_node() instead for
mapping "exception" debugfs file.

In running system the SOF_IPC4_DEBUG_SLOT_TELEMETRY is used for
telemetry data, but if configured, Zephyr exception data is dumped in
the same debug window slot, right after the separator word.

Signed-off-by: Jyri Sarha <jyri.sarha@linux.intel.com>

jsarha commented:
> @jsarha, what I would do is: rename the `ipc4-telemetry.c` to `ipc4-debugfs.c` or something generic and use this to export the raw slots on the needed base. You can have a single exported function as entry point, you can even move here the `sof_ipc4_find_debug_slot_offset_by_type()` function...
> 
> I know, mtrace is also uses debug slot, but it has been special cased out...
> 
> The `ipc4-telemetry.h` should remain as it is.

@ujfalusi I took a shot at this. I still try make sure the zephyr tool can still read the execption file.

jsarha commented:
@ujfalusi @kv2019i , according to my tests the exception debugfs file works just as well after this change as it did before. That is, neither version works. I have no idea where the problem is, but I am quite sure this version is equivalent to the older version, so I'd hope to get this merged.

lgirdwood commented:
@jsarha @ujfalusi ping - can we unblock ?

jsarha commented:
> @jsarha @ujfalusi ping - can we unblock ?

This is not as important anymore as before, as the latest debug_stream.py is able to access its debug-window-slot directly using cavstool.py functions. However, I still think the debugfs file is a cleaner way to do this, so I still think this PR is valid.

lgirdwood commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -56,16 +56,40 @@ sof_ipc4_sps_to_time_info(struct snd_sof_pcm_stream *sps)
+					      struct ipc4_pipeline_set_state_data *trigger_list)
+{
+	int i, offset = 0;
+
```
nitpick: I would always put a inline comment here on what is being constructed,

lgirdwood reviewed (commented):
LGTM.

lyakh commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -56,16 +56,40 @@ sof_ipc4_sps_to_time_info(struct snd_sof_pcm_stream *sps)
 {
 	struct sof_ipc4_msg msg = {{ 0 }};
 	u32 primary, ipc_size;
+	char debug_buf[32];
```
typical pipeline IDs have 1 or 2 digits, so 32 bytes should be enough for at least 10 IDs, right? Should be enough for now.

ujfalusi commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -56,16 +56,40 @@ sof_ipc4_sps_to_time_info(struct snd_sof_pcm_stream *sps)
 						   state);
+	}
+
+	dev_dbg(sdev->dev, "pipeline instances%s",
```
@jsarha, can you rebase and drop this patch?
We have this already, but in a cleaner way: f2f08c0471d5d

jsarha replied:
Oh I was hasty. I do not know how this ended up here. I must have rebased on some other branch than I intended to. Probably a local version of topic/sof-dev, instead of origin/topic/sof-dev. I'll drop the commit.

lyakh commented on sound/soc/sof/ipc4-debug-slot-debugfs.c:
```diff
@@ -0,0 +1,91 @@
+// This file is provided under a dual BSD/GPLv2 license.  When using or
+// redistributing this file, you may do so under either license.
+//
+// Copyright(c) 2024 Intel Corporation.
```
2025?

jsarha replied:
The file was created, the commit made, and the PR pushed to public github on 2024. Should the PRs always be edited to reflect the probable merge date? But I'll update it to 2024-2025 once we agree one the file and function names.

lyakh commented on sound/soc/sof/ipc4-debug-slot-debugfs.c:
```diff
@@ -0,0 +1,91 @@
+	}
+
+	*ppos = pos + count;
+	kfree(buf);
```
just move `kfree(buf);` to after line 53 and you don't need to call it twice

lyakh commented on sound/soc/sof/Makefile:
```diff
@@ -11,7 +11,7 @@ snd-sof-y +=	ipc3.o ipc3-loader.o ipc3-topology.o ipc3-control.o ipc3-pcm.o\
 ifneq ($(CONFIG_SND_SOC_SOF_IPC4),)
 snd-sof-y += ipc4.o ipc4-loader.o ipc4-topology.o ipc4-control.o ipc4-pcm.o\
-		ipc4-mtrace.o ipc4-telemetry.o
+		ipc4-mtrace.o ipc4-telemetry.o ipc4-debug-slot-debugfs.o
```
this implements read-only debugfs files, right? Should we add "ro" to the file name to make this clear?

jsarha replied:
If we ever need read-write access, wouldn't it be logical to extend this implementation without a need to rename the implementation file? How about changing function name sof_ipc4_create_debug_slot_debugfs_node() to sof_ipc4_create_debug_slot_ro_debugfs_node() ?

lyakh replied:
that's another possibility, yes

jsarha commented:
@lyakh is this Ok now?

@bardliao , @lgirdwood only appled two last suggestions from Guennadi (Copyright 2024-2025), and rename sof_ipc4_create_debug_slot_debugfs_node() to sof_ipc4_create_debug_slot_ro_debugfs_node() . Could you check those and reaprove if Ok.

jsarha commented:
@lgirdwood should we still get this merged? @ujfalusi should I rebase, the PR getting old, again?
